### PR TITLE
TST: fixtures for timezones

### PR DESCRIPTION
**Pull request by almaleksia**
This is a follow up for #20287 

Different tests use different lists of timezones and it's kind of hard to unify all of them. In this PR I tried to create a fixture that covers most frequent cases. Some tests use naive timezone, while others are designed to use only explicit timezones (and fail in case tz is not specified), that's why I created two fixtures to cover these two group of tests. List of timezones also includes one dateutil-timezone. 
 I'm not sure if described approach is correct though so I would appreciate any feedback.

- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/conftest.py [hunk 1]
```diff
@@ -97,3 +97,22 @@ def join_type(request):
...
     return request.param
+
+
+TIMEZONES = [None, 'UTC', 'US/Eastern', 'Asia/Tokyo', 'dateutil/US/Pacific']
+
+
+@pytest.fixture(params=TIMEZONES)
+def tz_naive(request):
+    """
+    Fixture for trying timezones including default (None)
```
Comment (jreback) [inline pandas/conftest.py]: can you make this ore explicit (e.g. list the timezones here); you can programatically construct the doc string I think (to make this DRY)

File: pandas/conftest.py [hunk 2]
```diff
@@ -97,3 +97,22 @@ def join_type(request):
...
+    """
+    return request.param
+
+
+@pytest.fixture(params=TIMEZONES[1::])
+def tz_aware(request):
+    """
+    Fixture for trying explicit timezones
```
Comment (jreback) [inline pandas/conftest.py]: same

File: pandas/conftest.py [hunk 3]
```diff
@@ -97,3 +97,22 @@ def join_type(request):
     Fixture for trying all types of join operations
     """
     return request.param
+
+
+TIMEZONES = [None, 'UTC', 'US/Eastern', 'Asia/Tokyo', 'dateutil/US/Pacific']
+
+
+@pytest.fixture(params=TIMEZONES)
+def tz_naive(request):
```
Comment (jreback) [inline pandas/conftest.py]: I wouldn't be averse to making the name of this slightly more explict

maybe ``tz_naive_fixture`` (maybe we should do this with the other fixtures as well, can be separate PR for that change)

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 4]
```diff
@@ -13,12 +13,17 @@
 from pandas.tests.test_base import Ops
 
 
+@pytest.fixture(params=[None, 'UTC', 'Asia/Tokyo', 'US/Eastern',
+                        'dateutil/Asia/Singapore',
+                        'dateutil/US/Pacific'])
+def tz(request):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: maybe ``tz_fixture``

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 5]
```diff
@@ -209,86 +210,85 @@ def test_nonunique_contains(self):
...
             assert idx[0] in idx
 
-    def test_order(self):
-        # with freq
-        idx1 = DatetimeIndex(['2011-01-01', '2011-01-02',
-                              '2011-01-03'], freq='D', name='idx')
-        idx2 = DatetimeIndex(['2011-01-01 09:00', '2011-01-01 10:00',
-                              '2011-01-01 11:00'], freq='H',
-                             tz='Asia/Tokyo', name='tzidx')
+    @pytest.mark.parametrize('idx',
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: you can format this more like
```
@pytest.mark.parameterize(
   'idx',
       [
         DatetimeIndex(......),
         DatetimeIndex(....)')
```

e.g. to fit more on a line

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@jbrockmendel ?

File: pandas/conftest.py [hunk 6]
```diff
@@ -97,3 +97,22 @@ def join_type(request):
...
+    """
+    Fixture for trying timezones including default (None)
+    """
+    return request.param
+
+
+@pytest.fixture(params=TIMEZONES[1::])
```
Comment (jbrockmendel) [inline pandas/conftest.py]: Is a single-colon OK here?

**Comment by almaleksia** [comment]
@jreback @jbrockmendel 
thank you for the feedback!

> I'm a bit concerned about always using the same ~4 timezones. Is there a risk of some special case weirdness in other timezones that we're missing?

Actually I had the same thought. As for this change, it affects only tests that already share similar list of timezones. I made sure that I didn't shorten or replace timezone lists used in the existing tests. Let's say test is using some timezone not covered by the fixture then list is not replaced. Several tests though got extended by one timezone (dateutil/US/Pacific) after being moved to the fixture. I thought about adding more timezones to the list but for some tests it looked a bit redundant, so I decided to save some build time.

> Would there be a use for separate fixtures for tzstr for tzinfo objects?

do you mean tests like in test_timezones parametrized with tzstr?

**Comment by TomAugspurger** [comment]
> I thought about adding more timezones to the list but for some tests it looked a bit redundant, so I decided to save some build time.

I think that's the right choice. This would be a decent case for using hypothesis: https://github.com/pandas-dev/pandas/issues/17978

File: pandas/conftest.py (hunk 1 above)
Comment (almaleksia) [reply pandas/conftest.py]: I added decorator for this

File: pandas/conftest.py (hunk 3 above)
Comment (almaleksia) [reply pandas/conftest.py]: Done :)

File: pandas/util/_test_decorators.py [hunk 7]
```diff
@@ -187,3 +187,10 @@ def decorated_func(func):
                                    "installed->{installed}".format(
                                        enabled=_USE_NUMEXPR,
                                        installed=_NUMEXPR_INSTALLED))
+
+
+def parametrize_fixture_doc(*args):
```
Comment (jreback) [inline pandas/util/_test_decorators.py]: this is nice, in a future PR can you document this itself?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
can you rebase this just to make sure

File: pandas/util/_test_decorators.py (hunk 7 above)
Comment (jreback) [reply pandas/util/_test_decorators.py]: can you add docs on this here

Comment (almaleksia) [reply pandas/util/_test_decorators.py]: @jreback done
sorry didn't see this comment at first, so I did two updates (one for rebase and one for doc), I think previous builds can be cancelled

**Comment by TomAugspurger** [comment]
Thanks @almaleksia!
